059-statically-configured-atc/proposal.md

```diff
@@ -0,0 +1,192 @@
+
+Running Concourse at scale, for a long period of time, poses a number of
+problems that make operability difficult.
+
+* In large organizations, it is not feasible for a team to set up Concourse by
+  themselves. If Concourse adoption is being driven by the engineering team
+  itself (i.e. bottom-up adoption), then the team is often challenged to
+  orchestrate the acquisition of servers (for `web` and for workers), storage, a
+  PostgreSQL database, networking, and credential access. If Concourse adoption
+  is being driven by engineering leadership (i.e. top-bottom adoption), then
+  leadership is challenged to orchestrate the synchronization of configuration
+  across a wide variety of Concourse instances that have been set up for
+  engineering teams.
+* For any sufficiently long-lived Concourse instance, database usage grows over
+  time, and the underlying PostgreSQL database grows vertically to match,
+  increasing the costs for end teams. Reducing reliance upon the underlying
+  PostgreSQL database has a direct, positive effect on instance performance and
+  costs incurred by Concourse administrators.
+* Perusing the current configuration for a Concourse instance, when so much of
+  the current configuration lives in the underlying PostgreSQL database, is
+  undocumented, difficult and error prone (if automation is attempted), and
+  changes are difficult to audit.
+* Configuring a `web` node is currently done either through passing flags at
+  runtime, or by setting environment variables prefixed with `CONCOURSE_`, as
+  [documented](https://concourse-ci.org/concourse-cli.html). While this means
+  of configuration is sufficient for providing one-line strings, e.g. setting a
+  connection string for a PostgreSQL database, it precludes the addition of more
+  complex multi-line configuration.
+
+# Proposal
+
+## Replacing Pre-Existing Command-Line Flags
+
+Permit running a `web` instance with a
+`--static-instance-configuration=/path/to/config.yaml` flag. If the
+`--static-instance-configuration` flag is provided, then no other command-line
+flags will be permitted; attempting to provide other flags (excepting the
+`--validate` flag, see below) will result in `concourse web` exiting with an
+error. Sending `SIGHUP` to the `web` process will ask the `web` process to
+reload this file, without restarting the process.
+
+The contents of this configuration YAML file will permit listing any of the
+existing options under `.options`. Sensitive strings may be substituted with
+`((FOO))` syntax, indicating that the string to be used is found in the
+environment variable `FOO`. For example (only a small subset is included,
+for brevity's sake):
+
+```yaml
+options:
+  external_url: https://concourse.example.com
+  enable_lets_encrypt: true
+  postgres_user: concourse
+  postgres_password: ((CONCOURSE_POSTGRES_PASSWORD))
+```
+
+## Replacing `fly set-team`
+
+Doing so permits the configuration file to include additional configuration that
```

Comment:
This is definitely an interesting idea, do you have a use case where an operator would like to restrict only being able to modify teams through the configuration file? My only concern with this is that it would require the operator to have to restart the web nodes everytime they want to make a change to the teams, but I do understand how it would be nice to have the true list of teams in the configuration file for the cluster. I guess I'm more curious of the "why" this would be nice for operators, to be able to be strict with the teams configuration?